fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/popcha/PoPCHAAuthenticationTest.java

```diff
@@ -0,0 +1,98 @@
+import com.github.dedis.popstellar.model.objects.security.Base64URLData;
+import com.github.dedis.popstellar.utility.security.Hash;
+import org.junit.Test;
+
+public class PoPCHAAuthenticationTest {
+
+  private static final String CLIENT_ID = Hash.hash("clientID");
+  private static final String NONCE = Hash.hash("random");
+  private static final Base64URLData IDENTIFIER = new Base64URLData("identifier");
+  private static final Base64URLData IDENTIFIER_PROOF = new Base64URLData("identifier-proof");
+  private static final String STATE = null;
+  private static final String RESPONSE_MODE = "id_token";
+  private static final String ADDRESS = "localhost:9100";
+
+  private static final PoPCHAAuthentication POPCHA_AUTHENTICATION =
+      new PoPCHAAuthentication(
+          CLIENT_ID, NONCE, IDENTIFIER, IDENTIFIER_PROOF, ADDRESS, STATE, RESPONSE_MODE);
+
+  @Test
+  public void getClientId() {
+    assertEquals(CLIENT_ID, POPCHA_AUTHENTICATION.getClientId());
+  }
+
+  @Test
+  public void getNonce() {
+    assertEquals(NONCE, POPCHA_AUTHENTICATION.getNonce());
+  }
+
+  @Test
+  public void getIdentifier() {
+    assertEquals(IDENTIFIER, POPCHA_AUTHENTICATION.getIdentifier());
+  }
+
+  @Test
+  public void getIdentifierProof() {
+    assertEquals(IDENTIFIER_PROOF, POPCHA_AUTHENTICATION.getIdentifierProof());
+  }
+
+  @Test
+  public void getState() {
+    assertEquals(STATE, POPCHA_AUTHENTICATION.getState());
+  }
+
+  @Test
+  public void getResponseMode() {
+    assertEquals(RESPONSE_MODE, POPCHA_AUTHENTICATION.getResponseMode());
+  }
+
+  @Test
+  public void getPopchaAddress() {
+    assertEquals(ADDRESS, POPCHA_AUTHENTICATION.getPopchaAddress());
+  }
+
+  @Test
+  public void getObject() {
+    assertEquals(Objects.POPCHA.getObject(), POPCHA_AUTHENTICATION.getObject());
+  }
+
+  @Test
+  public void getAction() {
+    assertEquals(Action.AUTH.getAction(), POPCHA_AUTHENTICATION.getAction());
+  }
+
+  @Test
+  public void testEquals() {
+    assertEquals(POPCHA_AUTHENTICATION, POPCHA_AUTHENTICATION);
+    assertNotEquals(null, POPCHA_AUTHENTICATION);
+
+    PoPCHAAuthentication popCHAAuthentication1 =
+        new PoPCHAAuthentication(
+            CLIENT_ID, NONCE, IDENTIFIER, IDENTIFIER_PROOF, ADDRESS, STATE, RESPONSE_MODE);
+    assertEquals(POPCHA_AUTHENTICATION, popCHAAuthentication1);
+  }
+
+  @Test
+  public void testHashCode() {
+    assertEquals(
+        java.util.Objects.hash(
+            CLIENT_ID, NONCE, IDENTIFIER, IDENTIFIER_PROOF, STATE, RESPONSE_MODE, ADDRESS),
+        POPCHA_AUTHENTICATION.hashCode());
+  }
+
+  @Test
+  public void testToString() {
+    String expected =
+        String.format(
+            "PoPCHAAuthentication{clientId='%s', nonce='%s', identifier='%s', identifierProof='%s', state='%s', responseMode='%s', popchaAddress='%s'}",
+            CLIENT_ID, NONCE, IDENTIFIER, IDENTIFIER_PROOF, STATE, RESPONSE_MODE, ADDRESS);
+    assertEquals(expected, POPCHA_AUTHENTICATION.toString());
+  }
+}
```

Comment:
and/or you could test this here as well.

Reply:
I'll add the non-happy tests for the PoPCHAQRCode object rather this (PoPPCHAAuthentication), since the latter is a message that the client itself craft and send over the network, so the non-happy path if it's the case is already intercepted before